src/utils.jl

```diff
@@ -97,15 +99,15 @@ gettypecodes(x::Dict{TypeCode, TypeCode}) = x
 # bit patterns for missing value, int value, escaped string, position and len in tape parsing
 const MISSING_BIT = 0x8000000000000000
 missingvalue(x::UInt64) = (x & MISSING_BIT) == MISSING_BIT
+sentinelvalue(::Type{Float64}) = Core.bitcast(UInt64, NaN) | MISSING_BIT
```

Comment:
One wrinkle I didn't anticipate from separating our previous `tape` into `tape` + `poslen` is how to encode `missing` values. Before we always had a `poslen` value around where we could steal a bit, but now we don't keep `poslen` around after parsing. It turns out that for all but `Int64`, we can find an empty bit pattern to use as a sentinel for `missing` (we use an obscure `NaN` bit pattern for `Float64`, and `Core.bitcast(UInt64, typemin(Int64))` for all other types).

I wrestled with what to do for `Int64`, since every bit pattern is a valid `Int64`. I first considered `typemin(Int64)`, but worried that it would turn out to actually be more frequent of a value encountered in the wild than a bunch of random `Int64`s (I'm thinking about manually encoded sentinels where they use `typemin(Int64)` themselves to mean something special). The approach here has 4 random Int64 values, positive and negative, and if we encounter one, we'll use the next one and recode the old sentinels. I think statistically we're in the clear here, but it is a bit unfortunate that this is so much uglier than other types.

Reply:
You mean that an error is thrown if the 4 values are encountered in the data? I guess that's OK, but OTC it would be a real problem if a missing value was silently generated instead of the actual value.

Reply:
Hmmm, actually we don't throw an error if we encounter all 4, but we probably should do a fatal error, or perhaps allow the user to manually pass a sentinel to use for Int.

Reply:
I'd throw an error for now and see whether that works. Providing a sentinel manually is really not user-friendly.

BTW, rather than random values, we can probably find good values by hand. For example, `typemax(Int)-1` is a really good candidate. OTC, anything close to 0 should be avoided.

Reply:
Why do we choose 4 sentinels. My intuition is that `INT_SENTINEL` could be a `Ref`. We initialize it somehow (here we can discuss). If we hit a collision (in practice I think it will be extremely unlikely) then we just carefully generate a new value to store in this `Ref` so that we are sure it was never encountered yet. In this way we will have no limie of 4 values, and moreover the code should be simpler (we do not need to keep track of the counter).

Reply:
Of course I might be missing something - I have just glanced over the code so please comment.

Reply:
And in this case this value should be local to a given column. An alternative would be to keep using the vector as you do and just grow it as needed (then it can be shared across in the single julia process).

Reply:
Here's how my thinking here has evolved, to catch everyone up:

* Ok, need to figure out a sentinel for `Int64`, R uses `typemin(Int64)`, so we could maybe do that
* But being `typemin(Int64)` might actually be too common if people hand encode things, so maybe `typemin(Int64) + 1`
* That's fine, but what if we happen to encounter that value while parsing? We could potentially just add 1 each time we encounter our sentinel, but that would be bad if they happen to have a sequence of values that correspond w/ our sequence of sentinels
* A BIG issue with changing the sentinel though, is that each time, in addition to going back and recoding past stored sentinel values to a new sentinel value, you're also worried about every past parsed value being equal to the new sentinel; I guess maybe that cost is worth simplifying things; i.e. if we happen to encounter our int sentinel, we pick a new Int64 sentinel randomly (not totally random, probably more like `rand(typemin(Int64):div(typemin(Int64), 10))`), and then, in addition to going back and recoding past sentinels, we also check that every parsed Int64 doesn't equal our new sentinel. Ultimately, that's not _that_ expensive of an operation in addition to having to recode our sentinel being very statistically unlikely.

Reply:
> (making sure to lock this vector in case multithreading is used).

So..........it turns out, I somehow completely didn't consider the multithreaded case in all of this and it's all obviously wrong (the recoding, bumping sentinels, and not just for Int types, but other types are affected as well). This probably explains why we still see issues in #495 and #484. I think I have an idea of how to adjust all this for multithreaded, so I'll take a stab at it.

Reply:
Thank you - please let me know when you are done so we can come back to this excellent PR.

Reply:
Ok, just pushed some new commits. The changes are these:

* Instead of using a single `AtomicVector` that each thread accesses/updates, we have a per-thread types vector
* So each thread will start "fresh" when parsing, detect values, and update/promote types as it's chunk of the file is parsed
* After all threads have finished, and similar to what we were already doing for pooled columns, we now "sync" the types detected from each thread and adjust if needed; e.g. if one thread parsed all Ints, it will have an Int column type, but another thread for the same column may have encountered a String value and promoted it. In our "sync" phase, we promote the types from each thread for each column, so that column's type would be promoted to String and we would recode the Int column values in that thread.
* The recoding/promoting of the "sync" phase isn't too bad/complex, and should be fairly rare, so I'm not too worried about any kind of serious performance hit.
* For the Int64 sentinel story, we now allocate an Int64 sentinel per column per thread, and do a similar "sync" afterwards, to ensure each column is encoded to a single sentinel value. It's a pain and a lot of code for how statistically unlikely it will be to ever hit that case, but whatever, we handle it and I just need to find a way to test it. Basically the logic is: do the per-thread parsing, afterwards, we set our `intsentinels` equal to the first thread's `intsentinels`, then check if any other column in any other thread ended up w/ a different int sentinel than the starting one (already pretty unlikely). If so, we then bump that column's int sentinel to the one that thread ended up (since we know it's valid for that thread). We then have a pass if a column's int sentinel isn't the starting one, to check if the new int sentinel is equal to any parsed value in any other thread. If so, we bump the int sentinel again and re-check all the threads until we find a sentinel that isn't present in the parsed value of any thread. We then do the actual recoding of each thread's sentinel values to the new, unique/good int sentinel for that column. Whew......

Reply:
I think that in order to test the `Int64` sentinel story we have to generate the data that contains exactly the sentinels that you are expecting to use (I have not dug in the code yet, but I assume that you are generating them in a way that can be made reproducible e.g. by setting appropriate seeds for PRNG).